dali/operators/reader/video_reader_decoder_op_test.cc

```diff
@@ -0,0 +1,299 @@
+      }
+
+      gt_frame_id += step;
+      ++sequence_id;
+
+      if (gt_frame_id + stride * sequence_length >= this->NumFrames(video_idx)) {
+        gt_frame_id = 0;
+        ++video_idx;
+        if (video_idx == this->NumVideos()) {
+          video_idx = 0;
+        }
+      }
+    }
+    ++batch_id;
+  }
+}
+
+TEST_F(VideoReaderDecoderTest, CpuVariableFrameRate) {
+  const int batch_size = 4;
+  const int sequence_length = 6;
+  const int stride = 3;
+  const int step = 10;
+
+  Pipeline pipe(batch_size, 4, 0);
+
+  pipe.AddOperator(OpSpec("experimental__readers__Video")
+    .AddArg("device", "cpu")
+    .AddArg("sequence_length", sequence_length)
+    .AddArg("stride", stride)
+    .AddArg("step", step)
+    .AddArg(
+      "filenames",
+      std::vector<std::string>{
+        testing::dali_extra_path() + "/db/video/vfr/test_1.mp4",
+        testing::dali_extra_path() + "/db/video/vfr/test_2.mp4"})
+    .AddArg("labels", std::vector<int>{0, 1})
+    .AddOutput("frames", "cpu")
+    .AddOutput("labels", "cpu"));
+
+  pipe.Build({{"frames", "cpu"}, {"labels", "cpu"}});
+
+  int num_sequences = 20;
+  int sequence_id = 0;
+  int batch_id = 0;
+  int gt_frame_id = 0;
+
+  int video_idx = 0;
+
+  while (sequence_id < num_sequences) {
+    DeviceWorkspace ws;
+    pipe.RunCPU();
+    pipe.RunGPU();
+    pipe.Outputs(&ws);
+
+    auto &frame_video_output = ws.template OutputRef<dali::CPUBackend>(0);
+    auto &frame_label_output = ws.template OutputRef<dali::CPUBackend>(1);
+
+    for (int sample_id = 0; sample_id < batch_size; ++sample_id) {
+      const auto sample = frame_video_output.tensor<uint8_t>(sample_id);
+      const auto label = frame_label_output.tensor<int>(sample_id);
+
+      ASSERT_EQ(label[0], video_idx);
+
+      for (int i = 0; i < sequence_length; ++i) {
+        this->ComapreFrames(
+          sample + i * this->FrameSize(video_idx),
+          this->GetVfrFrame(video_idx, gt_frame_id + i * stride),
+          this->FrameSize(video_idx));
+      }
+
+      gt_frame_id += step;
+      ++sequence_id;
+
+      if (gt_frame_id + stride * sequence_length >= this->NumFrames(video_idx)) {
+        gt_frame_id = 0;
+        ++video_idx;
+        if (video_idx == this->NumVideos()) {
+          video_idx = 0;
+        }
+      }
+    }
+    ++batch_id;
+  }
+}
+
+TEST_F(VideoReaderDecoderTest, RandomShuffle) {
+  const int batch_size = 1;
+  const int sequence_length = 1;
+
+  Pipeline pipe(batch_size, 1, 0, 1);
+
+  pipe.AddOperator(OpSpec("experimental__readers__Video")
+    .AddArg("device", "cpu")
+    .AddArg("sequence_length", sequence_length)
+    .AddArg("random_shuffle", true)
+    .AddArg("initial_fill", this->NumFrames(0))
+    .AddArg(
+      "filenames",
+      std::vector<std::string>{
+        testing::dali_extra_path() + "/db/video/cfr/test_1.mp4"})
+    .AddOutput("frames", "cpu"));
+
+  pipe.Build({{"frames", "cpu"}});
+
+  std::vector<int> expected_order = {44, 12, 49, 38, 8};
```

Comment:
why is that the expected order?

Reply:
I generated it locally and visually inspected the frames that we get for this seed. I put this test mostly to see, I this will be consistent among all builds we have. If no, I'll remove or modify this test.